### PR TITLE
feat: 🎸 add dev testing endpoint to aid with integration tests

### PR DESCRIPTION
**Pull request by polymath-eric**
### JIRA Link 

[DA-473](https://polymesh.atlassian.net/browse/DA-473)

### Changelog / Description 

adds `/developer-testing/create-admins` and
`/developer-testing/create-identity-batch` to allow for account setup during integration type tests

### Checklist - 

- [ ] New Feature ?
- [ ] Updated swagger annotation (if API structure is changed) ?
- [ ] Unit Test (if possible) ?
- [ ] Updated the Readme.md (if required) ?

[DA-473]: https://polymesh.atlassian.net/browse/DA-473?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DA-473]: https://polymesh.atlassian.net/browse/DA-473?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by polymath-eric** [comment]
Draft so the SDK beta into alpha can get merged (we need beta for subsidies, but alpha to clean up the `setSigner` call)

File: src/developer-testing/developer-testing.controller.ts [hunk 1]
```diff
@@ -27,4 +32,29 @@ export class DeveloperTestingController {
...
+      'Given a set of addresses this generates CDD claims, provides with a large amount of POLYX and makes them CDD providers',
+    description:
+      'This endpoint initializes a set of addresses to be chain admin accounts. The signer must best a CDD provider and have sufficient POLYX to cover the initial amounts (DEV ONLY)',
+  })
+  @Post('/create-admins')
+  async createAdmins(@Body() params: CreateCddProviders): Promise<IdentityModel[]> {
+    const ids = await this.developerTestingService.batchCreateAdmins(params);
+
+    return Promise.all(ids.map(id => createIdentityModel(id)));
+  }
```
Comment (prashantasdeveloper) [inline src/developer-testing/developer-testing.controller.ts:47]: @ApiResponse is missing here.

Comment (prashantasdeveloper) [reply src/developer-testing/developer-testing.controller.ts:47]: Also, just to be consistent I would recommend changing the return type to `Promise<ResultsModel<IdentityModel>>`. Can leave as it as well

File: src/developer-testing/developer-testing.controller.ts [hunk 2]
```diff
@@ -27,4 +32,29 @@ export class DeveloperTestingController {
...
+    summary: 'Creates a set of CDD claims for each address given',
+    description:
+      'This endpoint creates Identities for multiple accounts. Intended for testing purposes only',
+  })
+  @Post('/create-identity-batch')
+  async createIdentityBatch(@Body() params: CreateMockIdentityBatchDto): Promise<IdentityModel[]> {
+    const ids = await this.developerTestingService.batchCddClaimsWithSigner(params);
+
+    return Promise.all(ids.map(id => createIdentityModel(id)));
+  }
```
Comment (prashantasdeveloper) [inline src/developer-testing/developer-testing.controller.ts:59]: @ApiResponse is missing here as well

File: src/developer-testing/developer-testing.service.ts [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    const accounts = await this.fetchAccountsForAddresses(addresses);
+
+    return this.fetchAccountsIdentities(accounts);
+  }
+
+  private get alicePair(): KeyringPair {
+    if (!this._alicePair) {
+      const ss58Format = this.polymeshService.polymeshApi.network.getSs58Format().toNumber();
+      const keyring = new Keyring({ type: 'sr25519', ss58Format });
+      this._alicePair = keyring.addFromUri('//Alice');
```
Comment (prashantasdeveloper) [inline src/developer-testing/developer-testing.service.ts:152]: IMO we should add an extra variable in environments and default it to '//Alice'. That way if anyone wants to use this setup differently they don't have to edit the files

File: src/developer-testing/dto/create-admin.dto.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
+/* istanbul ignore file */
+
+import { ApiProperty } from '@nestjs/swagger';
+import { IsArray } from 'class-validator';
+
+export class CreateCddProviders {
```
Comment (prashantasdeveloper) [inline src/developer-testing/dto/create-admin.dto.ts:6]: nit -> Add Dto to the class name and make sure the class name and file name are in sync :)

File: src/developer-testing/developer-testing.service.ts [hunk 5]
```diff
@@ -0,0 +1,172 @@
+/* istanbul ignore file */
```
Comment (prashantasdeveloper) [inline src/developer-testing/developer-testing.service.ts:1]: Why are we ignoring the testing for this file? Atleast we should create a task to put in the tests later if we want this urgently. Not sure if I'm missing something here

File: src/developer-testing/developer-testing.service.ts [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+import { AccountsService } from '~/accounts/accounts.service';
+import { AppInternalError } from '~/common/errors';
+import { isNotNull } from '~/common/utils';
+import { CreateCddProviders } from '~/developer-testing/dto/create-admin.dto';
+import { CreateMockIdentityBatchDto } from '~/developer-testing/dto/create-mock-identity-batch';
+import { PolymeshService } from '~/polymesh/polymesh.service';
+import { SigningService } from '~/signing/services';
+
+const unitsPerPolyx = 1000000;
+const initialAdminPolyx = 100000000 * unitsPerPolyx;
```
Comment (prashantasdeveloper) [inline src/developer-testing/developer-testing.service.ts:18]: nit -> if we can provide an option to take the amount as input that will be really helpful. We can default to this value

File: src/developer-testing/dto/create-admin.dto.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+  @ApiProperty({
+    description: 'The addresses to create identities for and make CDD providers',
+    type: 'string',
+    isArray: true,
+    example: ['5GwwYnwCYcJ1Rkop35y7SDHAzbxrCkNUDD4YuCUJRPPXbvyV'],
+  })
+  @IsArray()
+  readonly addresses: string[];
+
```
Comment (prashantasdeveloper) [inline src/developer-testing/dto/create-admin.dto.ts:15]: nit-> We can add in amount / balance attribute to specify the initial POLYX to be transferred to each Identity. And can default it to the value used in service

**Review by prashantasdeveloper — CHANGES_REQUESTED** [review]
Left some nit comments, but overall LGTM

File: package.json [hunk 8]
```diff
@@ -8,6 +8,7 @@
   "scripts": {
     "nest": "nest",
     "prepare": "husky install",
+    "dedupe": "npx yarn-deduplicate yarn.lock",
```
Comment (polymath-eric) [inline package.json:11]: Updating the SDK introduced the warning message that happen all the time.

According to some SO post, yarn 2 is more aggressive in dedupe and will likely eliminate them for us.

For now, this package does the dedupe, but lets us stay on yarn 1 for now.

It only needs to be ran when you see the warnings after updating something touching polkadot dependencies.

**Comment by polymath-eric** [comment]
@prashantasdeveloper I refactored the old mock cdd endpoint to use the DeveloperTestingService.

I marked it as deprecated, since the batch call is better. I made that default to the sudo signer so its easier to migrate over, since that should be `//Alice`.

**Review by prashantasdeveloper — APPROVED** [review]
LGTM

**Pull request merged**